File: src/MatchingAlgorithm/LlcTopology.cpp
#include "LlcTopology.h"

#include <numbers>

double AngularFrequency(const double frequency)
{
	return 2 * std::numbers::pi * frequency;
}

LlcTopology::LlcTopology(HeatingSystem heatingSystem, const double inductance, const double capacitance)
	: Inductance(inductance), Capacitance(capacitance), _heatingSystem(std::move(heatingSystem))
{
	
}

double LlcTopology::Resistance(const double frequency, const double temperature) const
{
	const auto w = AngularFrequency(frequency);

	const auto resistance = _heatingSystem.Resistance(frequency, temperature);
	const auto inductance = _heatingSystem.Inductance(frequency, temperature);

	const auto req = resistance /
		(std::pow(1 - w * w * inductance * Capacitance, 2) +
			std::pow(w * Capacitance  * resistance, 2));

return req;
}

double LlcTopology::ParallelReactance(const double frequency, const double temperature) const
{
	const auto w = AngularFrequency(frequency);

	const auto resistance = _heatingSystem.Resistance(frequency, temperature);
	const auto inductance = _heatingSystem.Inductance(frequency, temperature);

	auto const xp = (w * inductance * (1 - w * w * inductance * Capacitance) - w * Capacitance * resistance * resistance) /
		(std::pow(1 * w * w * inductance * Capacitance, 2) + std::pow(w * Capacitance * resistance, 2));

	return xp;
}


double LlcTopology::Reactance(const double frequency, const double temperature) const
{
	const auto w = AngularFrequency(frequency);
	const auto x = w * Inductance * ParallelReactance(frequency, temperature);

	return x;
}


std::complex<double> LlcTopology::Impedance(const double frequency, const double temperature) const
{
	const auto resistance = Resistance(frequency, temperature);
	const auto reactance = Reactance(frequency, temperature);
	
	auto const z = std::complex(resistance, reactance);

	return z;
	
}

File: src/MatchingAlgorithm/Llc/LlcTopology.cs
using System.Numerics;

namespace MatchingAlgorithm.Llc;

public class LlcTopology : ILlcTopology
{
    private readonly IHeatingSystem _heatingSystem;

    private readonly Func<double, double, double> _ihsResistance;
    private readonly Func<double, double, double> _ihsReactance;

    public LlcTopology(IHeatingSystem heatingSystem)
    {
        _heatingSystem = heatingSystem;
        _ihsResistance = heatingSystem.Resistance;
        _ihsReactance = heatingSystem.Reactance;
    }

    public double Resistance(double frequency, double temperature)
    {
        var w = AngularFrequency(frequency);
        var r = _heatingSystem.Resistance(frequency, temperature);
        var l = _heatingSystem.Reactance(frequency, temperature);
        
        var req = r /
                  (Math.Pow(1 - w * w * l * Capacitance, 2) +
                   Math.Pow(w * Capacitance * r, 2));

        return req;
    }

    public double Reactance(double frequency, double temperature)
    {
        return AngularFrequency(frequency) * Inductance + ParallelReactance(frequency, temperature);
    }

    public Complex Impedance(double frequency, double temperature)
    {
        return new Complex(Resistance(frequency, temperature), Reactance(frequency, temperature));
    }

    public double Inductance { get; set; }
    public double Capacitance { get; set; }

    public double ParallelReactance(double frequency, double temperature)
    {
        var w = AngularFrequency(frequency);
        var r = _heatingSystem.Resistance(frequency, temperature);
        var l = _heatingSystem.Reactance(frequency, temperature);
        
        var xp =
            (w * l * (1 - w * w * l * Capacitance) - w * Capacitance * r * r) /
            (Math.Pow(1 - w * w * l * Capacitance, 2) + Math.Pow(w * Capacitance * r, 2));

        return xp;
    }

    private static double AngularFrequency(double frequency)
    {
        return 2 * Math.PI * frequency;
    }
}